drivers/entropy/entropy_nrf5.c

```diff
@@ -8,92 +8,253 @@
 #include <entropy.h>
 #include <atomic.h>
 #include <nrf_rng.h>
+#include <soc.h>
+
+struct rand {
+	u8_t count;
+	u8_t threshold;
+	u8_t first;
+	u8_t last;
+	u8_t rand[0];
+};
+
+#define RAND_DEFINE(name, len) u8_t name[sizeof(struct rand) + len] __aligned(4)
+
+#define RAND_THREAD_LEN (CONFIG_ENTROPY_NRF5_THR_BUF_LEN + 1)
+#define RAND_ISR_LEN    (CONFIG_ENTROPY_NRF5_ISR_BUF_LEN + 1)
 
 struct entropy_nrf5_dev_data {
-	atomic_t user_count;
+	struct k_sem sem_lock;
+	struct k_sem sem_sync;
+
+	RAND_DEFINE(thr, RAND_THREAD_LEN);
+	RAND_DEFINE(isr, RAND_ISR_LEN);
 };
 
 #define DEV_DATA(dev) \
 	((struct entropy_nrf5_dev_data *)(dev)->driver_data)
 
-static inline u8_t entropy_nrf5_get_u8(void)
+static size_t get(struct rand *rng, size_t octets, u8_t *rand)
 {
-	while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
-		__WFE();
-		__SEV();
-		__WFE();
+	u8_t first, last, remaining;
+
+	__ASSERT_NO_MSG(rng);
+
+	first = rng->first;
+	last = rng->last;
+
+	if (first <= last) {
+		u8_t *d, *s;
+		u8_t avail;
+
+		d = &rand[octets];
+		s = &rng->rand[first];
+
+		avail = last - first;
+		if (octets < avail) {
+			remaining = avail - octets;
+			avail = octets;
+		} else {
+			remaining = 0;
+		}
+
+		first += avail;
+		octets -= avail;
+
+		while (avail--) {
+			*(--d) = *s++;
+		}
+
+		rng->first = first;
+	} else {
+		u8_t *d, *s;
+		u8_t avail;
+
+		d = &rand[octets];
+		s = &rng->rand[first];
+
+		avail = rng->count - first;
```

Comment:
can we have a comment here? - like `fetch bytes form upper part of random buffer.`